pkg/sql/zoneconfig/zone_config_options.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package zoneconfig
+
+import (
+	"context"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/pkg/config/zonepb"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"google.golang.org/protobuf/proto"
+	"gopkg.in/yaml.v2"
+)
+
+// EnterpriseCheckCallback call back for validating if a feature requires an
+// enterprise license.
+type EnterpriseCheckCallback func(featureName string) error
+
+// SupportedZoneConfigOptions indicates how to translate SQL variable
+// assignments in ALTER CONFIGURE ZONE to assignments to the member
+// fields of zonepb.ZoneConfig.
+var SupportedZoneConfigOptions = map[tree.Name]struct {
+	RequiredType *types.T
+	Setter       func(*zonepb.ZoneConfig, tree.Datum)
+	Getter       func(*zonepb.ZoneConfig) (tree.Datum, error)
+	CheckAllowed func(context.Context, EnterpriseCheckCallback, tree.Datum) error // optional
```

Comment:
nit: how would you feel about making this type exported and have these fields as private fields and then have methods which call them? The `CheckAllowed` could then no-op if the optional field is not set.

For testing you can expose a constructor. Then, lastly, unexport the map and instead have a public function to do the map lookup